platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/cache/ConcurrentArray.java

```diff
@@ -134,49 +143,48 @@
         }
 
         // Add the first array, so we are primed and ready to go. This also saves on a check later for an empty
-        // concurrent linked dequeue (we always know there is at least one sub-array).
+        // linked list (we always know there is at least one sub-array).
         this.subarrayCapacity = capacity;
-        arrays.add(new SubArray<>(capacity));
+        current = new AtomicReference<>();
+        current.set(new SubArray<>(capacity));
     }
 
     /**
-     * Create a new {@link ConcurrentArray} with the contents of two other {@link ConcurrentArray}s.
-     * <strong>IMPORTANT!</strong> this does <strong>NOT</strong> copy the array data out of the source
-     * {@link ConcurrentArray}s. It directly reuses the data. Both of the source arrays must be immutable.
-     * This array will be <strong>sealed</strong>. It cannot be mutated after merging.
-     *
-     * @param arrayOne
-     * 		The first array to take data from. Cannot be null. Must be immutable.
-     * @param arrayTwo
-     * 		The second array to take data from. Cannot be null. Must be immutable.
-     * @throws NullPointerException
-     * 		if either array is null
-     * @throws IllegalArgumentException
-     * 		if either array is mutable, or if they are the same instance.
+     * Create a new copy from the current one. We can't set {@code head} in the constructor.
+     * @param other
+     *      The previous copy sharing the linked list of {@link SubArray}s.
      */
-    ConcurrentArray(ConcurrentArray<T> arrayOne, ConcurrentArray<T> arrayTwo) {
-        Objects.requireNonNull(arrayOne);
-        Objects.requireNonNull(arrayTwo);
+    ConcurrentArray(final ConcurrentArray<T> other) {
+        this.subarrayCapacity = other.subarrayCapacity;
+        this.current = other.current;
+        this.head = null;
+    }
+
+    /**
+     * Effectively add all {@link SubArray}s from the previous copy.
+     * @param other
+     *      The previous copy from the same family.
+     */
+    void merge(final ConcurrentArray<T> other) {
+        Objects.requireNonNull(other);
 
         // We don't allow either to be mutable
-        if (!arrayOne.isImmutable() || !arrayTwo.isImmutable()) {
-            throw new IllegalArgumentException("The source arrays *must* be immutable");
+        if (!this.isImmutable() || !other.isImmutable()) {
```

Comment:
Consider rewriting this as `if(!(this.isImmutable() && other.isImmutable()))`  It reads out closer to what the exception message says is required, and might be easier for the next engineer to follow.

Reply:
Done